src/serialize/json/denormalize.ts

```diff
@@ -0,0 +1,452 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+SPDX-License-Identifier: Apache-2.0
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+import { PackageURL } from 'packageurl-js'
+import { URL } from 'url'
+
+import { ComponentScope, HashAlgorithm } from '../../enums'
+import * as Models from '../../models'
+import type { Protocol, Protocol as Spec } from '../../spec'
+import { Format, SpecVersionDict, UnsupportedFormatError } from '../../spec'
+import type { DenormalizerOptions } from '../types'
+import type { Normalized } from './types'
+
+type VarType = 'string' | 'number' | 'bigint' | 'boolean' | 'symbol' | 'undefined' | 'object' | 'function'
+type PathPart = string | number
+type PathType = PathPart[]
+
+function formatPath (path: PathType): string {
+  return path.map(p => typeof p === 'number' ? `[${p}]` : `.${p}`).join('')
+}
+
+function assertTypes (value: any, expected: VarType[], path: PathType): void {
+  const is = typeof value
+  if (!expected.includes(is)) {
+    throw new TypeError(`${formatPath(path)} is ${is} but should be one of ${expected.join(', ')}`)
+  }
+}
+
+function assertNonEmptyStr (value: any, path: PathType): asserts value is string {
+  assertTypes(value, ['string'], path)
+  if (value.length === 0) {
+    throw new RangeError(`${formatPath(path)} should be non empty string`)
+  }
+}
+
+function checkEnum (value: any, allowed: any[]): boolean {
+  return allowed.every(ev => ev !== value)
+}
+
+function assertEnum (value: any, allowed: any[], path: PathType): void {
+  if (!checkEnum(value, allowed)) {
+    throw new TypeError(`${formatPath(path)} is ${JSON.stringify(value)} but should be one of ${JSON.stringify(allowed)}`)
+  }
+}
+
+function assertArrayOrNull (value: any, path: PathType): asserts value is (any[] | undefined) {
+  if (value != null && !Array.isArray(value)) {
+    throw new TypeError(`${formatPath(path)} is ${typeof value} but should be an array or undefined`)
+  }
+}
+
+function captureErrorInPath<T> (func: () => T, path: PathType): T {
+  try {
+    return func()
+  } catch (e) {
+    if (e instanceof Error) {
+      e.message = `Error in ${formatPath(path)}: ${e.message}`
+      throw e
+    } else {
+      throw new Error(`Unknown error in ${formatPath(path)}`)
+    }
+  }
+}
+
+function createRepository<VT, RT> (
+  arr: any,
+  options: DenormalizerOptions,
+  path: PathType,
+  denormalizer: BaseJsonDenormalizer<VT, any>,
+  Repository: new(arr: VT[]) => RT
+): RT | undefined {
+  assertArrayOrNull(arr, path)
+  return (arr != null)
+    ? new Repository(arr.map((item: any, index) => denormalizer.denormalize(item, options, [...path, index])))
+    : undefined
+}
+
+export class Factory {
+  #spec?: Spec
```

Comment:
:question: the fact that this is not `readonly` makes me wonder what would happen,
if I used this class in a parallel thread? 
I mean, JavaScript is all about async and event based and in general parallel.

That would happen if i used the same object with two different deserializes on two different BOMs?
They would change the `#spec` sight on use, which might result in unpredictable results. 
Please have this fixed. See https://github.com/CycloneDX/cyclonedx-javascript-library/pull/506/files#r1115847206